c10/util/intrusive_ptr.h

```diff
@@ -33,6 +33,8 @@ constexpr uint64_t kImpracticallyHugeWeakReferenceCount =
 constexpr uint64_t kReferenceCountOne = 1;
 constexpr uint64_t kWeakReferenceCountOne = (kReferenceCountOne << 32);
 constexpr uint64_t kUniqueRef = (kReferenceCountOne | kWeakReferenceCountOne);
+// Indicates whether the object has a PyObject wrapper.
+constexpr uint64_t kHasPyObject = (uint64_t(1) << 63);
```

Comment:
can/should we define a compile-time check to rule out some types T such that `intrusive_ptr<T>` cannot possibly have a pyobject and 1) save the extra instructions for those types and 2) check this compile-time assumption (ideally at compile time, but runtime is OK) by failing if the new pyobject stuff is called? I imagine that those types exist (do we ever back e.g. c10::List with a PyObject?) and would at least appreciate the code size savings

Reply:
oops, I realize now that I should have suggested that instead we allowlist those types that require PyObject support rather than ruling out specific types.

Reply:
> But in a few places we use the generic intrusive_ptr<intrusive_ptr_target> and I think that would end up (incorrectly) ignoring the PyObject wrappers when the actual target is a Tensor/Storage.

Couldn't we solve that by opting intrusive_ptr_target into the has_pyobject behavior? this shouldn't affect subclasses; instantiating TargetTraits<T> shouldn't care about inheritance